### ISSUE TITLE
New BIDS file requires testing on pipeline steps 

### ISSUE THREAD

**Issue by F-said** [NDCLab/baseEEG#59]
BIDS file created from dev-feature-io should work on:

- feature-filter
- feature-segment

**Comment by F-said** [comment]
## feature-filter

Running: 

```python
bids_root = pathlib.Path('BIDS')
bids_path = mne_bids.BIDSPath(subject='NDARAB793GL3',
                              session='01',
                              task='ContrastChangeBlock1',
                              run='01',
                              datatype='eeg',
                              root=bids_root)
raw = mne_bids.read_raw_bids(bids_path)

raw_plot = raw.plot()

# show and close 
matplotlib.pyplot.show()
matplotlib.pyplot.close()

# filter
raw_filtered = baseeeg_preprocessing_filter(raw)

raw_filtered = raw_filtered.plot()

# show and close 
matplotlib.pyplot.show()
matplotlib.pyplot.close()
```

Successfully produces filtered and unfiltered raw data.

Graphs created as well: 

### Unfiltered
![image](https://user-images.githubusercontent.com/26397102/115919615-af3efb00-a43e-11eb-921c-71e9a3b81687.png)

### Filtered
![image](https://user-images.githubusercontent.com/26397102/115919684-c8e04280-a43e-11eb-8647-a80072f0a6db.png)

Light runtime warning:
```
RuntimeWarning: Unable to map 'hand' value to MNE. Not setting subject handedness.
```

Original filter file instead mapped 'left-hand' and 'right-hand' info. Is this ok to ignore? @georgebuzzell. 

## feature-segment 

Running:

```python
# extract events and event_id from our raw object
events, event_id = mne.events_from_annotations(raw)

# we can go ahead and plot the events
mne.viz.plot_events(events, event_id=event_id)

# create an epoch object and we can extract information off of it
epochs = mne.Epochs(raw, events, event_id=event_id, preload=True)
handle = epochs.plot(n_epochs=10)
matplotlib.pyplot.show()
matplotlib.pyplot.close()
print(epochs.info)
```

Successfully produced epochs based on events with the following annotations:
```
Used Annotations descriptions: ['12', '5', '8', '82', '9', '9999']
Not setting metadata
Not setting metadata
72 matching events found
Setting baseline interval to [-0.2, 0.0] sec
Applying baseline correction (mode: mean)
0 projection items activated
Loading data for 72 events and 351 original time points ...
1 bad epochs dropped
<Info | 10 non-empty values
 bads: []
 ch_names: E1, E2, E3, E4, E5, E6, E7, E8, E9, E10, E11, E12, E13, E14, ...
 chs: 129 EEG
 custom_ref_applied: False
 dig: 129 items (129 EEG)
 highpass: 0.0 Hz
 line_freq: 60.0
 lowpass: 250.0 Hz
 meas_date: 2017-05-12 16:07:18 UTC
 nchan: 129
 projs: []
 sfreq: 500.0 Hz
 subject_info: 5 items (dict)
>
```

Along with visual:

### Events
![image](https://user-images.githubusercontent.com/26397102/115922086-0eead580-a442-11eb-81b0-e9c71b780039.png)

### Epochs
![image](https://user-images.githubusercontent.com/26397102/115923181-79e8dc00-a443-11eb-802e-c2d665cd3f33.png)

BIDS from feature-io works with previously created code! @DMRoberts @georgebuzzell

**Comment by georgebuzzell** [comment]
Woo hoo! This is great!! Thank you for testing this, @F-said !

I think it is totally fine to ignore the hand mapping for the forseeable future. I can't imagine us actually needing mne to know that. It is good to have that data, but it is fine for that data to exist elsewhere.

**Comment by georgebuzzell** [comment]
@F-said or @DMRoberts can one of you two please add the bids-formatted output of feature-io here: https://drive.google.com/drive/u/0/folders/1aQY97T9EfkPEkuiCav2ei9cs0DFegO4-

Then, can you please create a pull request editing the roadmap readme to indicate that everyone should use that file for development (and include the link to the file)? Please link the PR to this issue.

THANK YOU!!

**Comment by yanbin-niu** [comment]
@georgebuzzell I got the reref feature tested with our BIDS file:

```
bids_root = pathlib.Path('BIDS')
bids_path = mne_bids.BIDSPath(subject='NDARAB793GL3',
                              session='01',
                              task='ContrastChangeBlock1',
                              run='01',
                              datatype='eeg',
                              root=bids_root)
raw = mne_bids.read_raw_bids(bids_path)
raw.plot()

# average referencing
raw_avg_ref = baseeeg_preprocessing_reref(raw, ref_channels=['FZ'])
raw_avg_ref.plot()
```

Graphs before vs after:
![image](https://user-images.githubusercontent.com/79607547/116048896-42c73480-a6a8-11eb-9a1d-633165326975.png)
![image](https://user-images.githubusercontent.com/79607547/116048941-507cba00-a6a8-11eb-8ea3-f03f25c7525a.png)

Log:
![image](https://user-images.githubusercontent.com/79607547/116049141-8a4dc080-a6a8-11eb-93e6-5c2d66eb73fa.png)

Sorry, I still did not figure out the plotting issue in my native environment. So, I just plotted in the container and it can not plot in a GUI. But, we can see there are some changes, though hard to see....... Along with the log, it should be good.


Some questions:
1. do we just have to test the function with one task, since we have multiple tasks?
2. I get a runtime warning, when reading the BIDS file:

> RuntimeWarning: Limited 1 annotation(s) that were expanding outside the data range.
>   raw = mne_bids.read_raw_bids(bids_path)

3. I got another runtime waring:

> <ipython-input-2-c8f254504aee>:9: RuntimeWarning: The locations of multiple reference channels are ignored (set to zero).
>   raw_new_ref = mne.add_reference_channels(raw, ref_channels=ref_channels)

I thought it's related to adding back the online reference channel. While working with sample data, we added FZ back. My question is , for our data, which online reference electrode should be added back or we do not have to add back any electrode.

**Comment by georgebuzzell** [comment]
@yanbin-niu this is great!! answers to your questions:

No need to test on multiple tasks for this feature. But a great question!

The runtime warning for an annotation expanding out of range does not sound like a problem. I believe what is happening is this: when you segment/epoch your data, you are asking mne to cut a segment of x length around the annotation time point. However, if the annotation occurs near the very beginning or very end of the continuous recording, then when you go to segment the data, there will not be enough data to cut size x around the annotation. Hence, the warning. At least, that is my understanding. Can you please confirm that the runtime warning occurs when epoching? Or does, it happen when you simply read in the data?

For the runtime warning for rereferencing: It looks like you are specifying Fz as the reference in your code right now. The reference for these data should actually be Cz I think. It should already be in the data, so you do not need to add it back in. I think the warning you are getting is because you are telling mne that fz is the reference, but really, cz is the reference and mne is saying that might be the case because cz is all zeros. can you please confirm?

**Comment by yanbin-niu** [comment]
@georgebuzzell 
The runtime warning happened when I simply read in the data. Dan may have some great insights on this.

I double checked the Cz channel. As you said, it is all zeros.
<img width="521" alt="image" src="https://user-images.githubusercontent.com/79607547/116261986-bf423c00-a7aa-11eb-8442-141c2fa728e4.png">

Also, if I do not add back any channel, there will be no warning.
<img width="698" alt="image" src="https://user-images.githubusercontent.com/79607547/116262438-1c3df200-a7ab-11eb-9441-45b65039061a.png">

It looks good now. Let me know if I should do any update on the script.

**Comment by georgebuzzell** [comment]
@yanbin-niu Perfect! Yes, please update the script as you describe. Please do a PR for feature-filter with this change. in your PR please note that the runtime warning for annotations expanding outside the range as unrelated to the feature-filter test. 

We do need to resolve the runtime warning for annotations, but that is not related to feature-filter. Please post a seperate issue and @DMRoberts and @F-said for that issue.

**Comment by yanbin-niu** [comment]
@georgebuzzell I think we might not need to update the script for now, since I assume we may want to read the online reference channel (either None or a channel name) from .json file. The current function is capable of receiving this parameter:

```
def baseeeg_preprocessing_reref(raw: mne.io.Raw, ref_channels=None) -> mne.io.Raw:
    try:
        raw.load_data()
        
        # add back reference channel (all zero)
        if ref_channels is None:
            raw_new_ref = raw
        else:
            raw_new_ref = mne.add_reference_channels(raw, ref_channels=ref_channels)

        # return average reference
        return raw_new_ref.copy().set_eeg_reference(ref_channels='average')
    
    except TypeError:
        print('Type Error')
        
    except Exception:
        print('Unknown Error')
```

Does this make sense? (btw, this is for the re-reference feature)

I will make an issue about the runtime warning.

**Comment by georgebuzzell** [comment]
@yanbin-niu ok, great! Agreed!

**Comment by DMRoberts** [comment]
@georgebuzzell @yanbin-niu That's correct that Cz is the reference for the example data. What is a little odd is that we are writing the reference to *eeg.json when we write the BIDS dataset, ie:

`"EEGReference": "Cz"`

However it seems that maybe MNE isn't using that information when it reads the BIDS dataset in? I was also surprised when writing the BIDS dataset, that even if I ran the 'set_eeg_reference' function specifying that the Cz electrode is the reference, I still had to manually specify that Cz  was the reference to the BIDS exporting function - 'setting' the reference didn't seem to store a label that the reference was set to Cz. This actually seems to be something that is lacking in MNE at the moment as noted in the following open issue with MNE: https://github.com/mne-tools/mne-python/issues/8962

**Comment by georgebuzzell** [comment]
Thank you, Dan!

@F-said can ubplease create an issue that links to the mne issue dan posted above. That way, we can track and keep an eye on this, then update our code in the future as needes?

@yanbin-niu I think we are fine to move forward though as is.

@DMRoberts what about the annotations expanding warning?

**Comment by F-said** [comment]
resolved with merged PR #90
